### PR TITLE
Teamcity webhook

### PR DESCRIPTION
**Pull request by firephreek**
Implements a basic TeamCity Webhook.

TC Webhooks require:
- HostUrl
- Auth Token
- VCS Root Id/Name

TC Hook is just a REST API that takes a POST with no body and an "Authorization: Bearer <token>" Header.

Translations properties added, templates created, logo included.

Go isn't my primary or even secondary language, so happy to hear changes that would be needed.
Adding the BearerToken property to the HookTask seemed the least problematic approach to getting the AuthToken included in the final POST, but if there's a more 'GO' approach or something specific to the project, I don't know it.

### CODE REVIEW

**Comment by lunny** [comment]
Hi, you have an outdate code base.

File: modules/webhook/teamcity.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package webhook
```
Comment (techknowlogick) [inline modules/webhook/teamcity.go]: Please add copyright header

File: modules/webhook/teamcity.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"encoding/json"
```
Comment (techknowlogick) [inline modules/webhook/teamcity.go]: Please split up imports with built in first, then an empty line before code.Gitea.io ones

File: options/locale/locale_fa-IR.ini [hunk 3]
```diff
@@ -1483,9 +1483,13 @@ settings.add_slack_hook_desc=درست کردن <a href="%s">سستی </a> در 
 settings.slack_token=توکن
 settings.slack_domain=دامنه
 settings.slack_channel=کانال
+settings.teamcity_host_url = Host URL
+settings.teamcity_vcs_root_id = VCS Root ID
```
Comment (techknowlogick) [inline options/locale/locale_fa-IR.ini]: Please only change English translations as our translation system will overwrite any that aren't in its system

File: templates/swagger/v1_json.tmpl [hunk 4]
```diff
@@ -11687,6 +11687,7 @@
             "gitea",
             "gogs",
             "msteams",
+						"teamcity",
```
Comment (techknowlogick) [inline templates/swagger/v1_json.tmpl]: Please ensure indentation is the same on this line as the others

**Review by techknowlogick — COMMENTED** [review]
Haven't taken a look at the code, but there are a few nitpicks that the CI will likely fail on.

Thanks for the PR :)

File: templates/admin/hook_new.tmpl [hunk 5]
```diff
@@ -22,6 +22,8 @@
 					<img class="img-13" src="{{StaticUrlPrefix}}/img/dingtalk.ico">
 				{{else if eq .HookType "telegram"}}
 					<img class="img-13" src="{{StaticUrlPrefix}}/img/telegram.png">
+				{{else if eq .HookType "teamcity"}}
+					<img class="img-13" src="{{StaticUrlPrefix}}/img/teamcity.png">
```
Comment (silverwind) [inline templates/admin/hook_new.tmpl]: There is a SVG version available: https://commons.wikimedia.org/wiki/File:TeamCity_Icon.svg

File: modules/webhook/teamcity.go (hunk 1 above)
Comment (firephreek) [reply modules/webhook/teamcity.go]: ✔

File: modules/webhook/teamcity.go (hunk 2 above)
Comment (firephreek) [reply modules/webhook/teamcity.go]: ✔

File: options/locale/locale_fa-IR.ini (hunk 3 above)
Comment (firephreek) [reply options/locale/locale_fa-IR.ini]: ✔

File: templates/swagger/v1_json.tmpl (hunk 4 above)
Comment (firephreek) [reply templates/swagger/v1_json.tmpl]: ✔

File: templates/admin/hook_new.tmpl (hunk 5 above)
Comment (firephreek) [reply templates/admin/hook_new.tmpl]: ✔

**Comment by firephreek** [comment]
PR Rebased. Thought I was at latest but upstream wasn't pulling correctly, should be resolved. 
Added a small unit test for the webhook, just checks that the payload isn't altered and that the Meta content is marshalled correctly.

**Comment by KN4CK3R** [comment]
This contains changes which should not be part of this PR.

**Comment by firephreek** [comment]
> This contains changes which should not be part of this PR.

Found and reverted the offending files. I think they got picked up when I rebased.

File: templates/admin/hook_new.tmpl [hunk 6]
```diff
@@ -26,8 +26,12 @@
 					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/dingtalk.ico">
 				{{else if eq .HookType "telegram"}}
 					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/telegram.png">
+				{{else if eq .HookType "teamcity"}}
+					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/teamcity.svg">
 				{{else if eq .HookType "msteams"}}
 					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/msteams.png">
+				{{else if eq .HookType "teamcity"}}
+					<img width="26" height="26" src="{{AssetUrlPrefix}}/img/teamcity.png">
```
Comment (silverwind) [inline templates/admin/hook_new.tmpl]: Duplicate image?

File: templates/org/settings/hook_new.tmpl [hunk 7]
```diff
@@ -21,8 +21,12 @@
 							<img width="26" height="26" src="{{AssetUrlPrefix}}/img/dingtalk.ico">
 						{{else if eq .HookType "telegram"}}
 							<img width="26" height="26" src="{{AssetUrlPrefix}}/img/telegram.png">
+						{{else if eq .HookType "teamcity"}}
+							<img width="26" height="26" src="{{AssetUrlPrefix}}/img/teamcity.svg">
 						{{else if eq .HookType "msteams"}}
 							<img width="26" height="26" src="{{AssetUrlPrefix}}/img/msteams.png">
+						{{else if eq .HookType "teamcity"}}
+							<img width="26" height="26" src="{{AssetUrlPrefix}}/img/teamcity.png">
```
Comment (silverwind) [inline templates/org/settings/hook_new.tmpl]: duplicate?

File: templates/org/settings/hook_new.tmpl [hunk 8]
```diff
@@ -41,7 +45,9 @@
 					{{template "repo/settings/webhook/discord" .}}
 					{{template "repo/settings/webhook/dingtalk" .}}
 					{{template "repo/settings/webhook/telegram" .}}
+					{{template "repo/settings/webhook/teamcity" .}}
 					{{template "repo/settings/webhook/msteams" .}}
+					{{template "repo/settings/webhook/teamcity" .}}
```
Comment (silverwind) [inline templates/org/settings/hook_new.tmpl:47]: duplicate?

Comment (firephreek) [reply templates/org/settings/hook_new.tmpl:47]: I hate rebasing so much sometimes....resolved!  Thanks for the catch!

**Comment by zeripath** [comment]
This branch is a total mess. You've reverted lots of important things and caused loads of conflicts. 

Do you have a SHA from before you attempted to rebase?

**Review by zeripath — CHANGES_REQUESTED** [review]
There are lots of serious reversions here due to mismerges. This cannot merge as is.

**Comment by wxiaoguang** [comment]
> any news on this issue?

This PR's merge history is still broken.

And in future I think all webhooks could be covered by a general mechanism, instead of hard-code more and more webhooks.
* #19307

**Comment by techknowlogick** [comment]
closing per above comment

**Pull request closed without merging**